### ISSUE TITLE
Rebinding hjkl for different keyboard layouts

### ISSUE THREAD

**Issue by niklasfasching** [syl20bnr/spacemacs#1391]
This is obviously a fringe case, but how would one go about rebinding hjkl to a different set of keys everywhere? I'm using neo-layout.org and hjkl is not usable (see #1255). 

I am rebinding the normal keymaps like this
```elisp
;; single key bindings have to be changed per mode
(define-key evil-normal-state-map "s" 'evil-backward-char)
(define-key evil-normal-state-map "h" 'evil-substitute) ;s

;; bindings with modifier should be safe to be changed globally
(keyboard-translate ?\C-h ?\C-s)
(keyboard-translate ?\C-s ?\C-h)


```
but am stuck at redefining micro-state keymaps to change the bindings in e.g. helm. Is it possible to change a micro-state keymap?
Also, is there a keyboard-translate that would work for single letters (only switching the keys when looking up functions, not when typing normally)?

Thanks

**Comment by syl20bnr** [comment]
There is no support in micro-state right now to extend them. Doing what you propose is opening the pandora box, you may end with a lof of shadowed key bindings and any update of spacemacs can become an headache.
Looking at the neo layout I suggest you to just remap `j`.

**Comment by Nupf** [comment]
You're right. I was hoping I could avoid that by switching keys whenever they are bound to a function (so "h" acts as "s" in bindings, but normally when typing (not in a binding)
Sadly I haven't found a solution (maybe I'm trying something impossibly complicated here and that's why I'm not finding anything - please tell me in that case, I'm pretty new to emacs).

I was thinking I could translate the keys globally and then bind them back when not shadowed by a binding
```elisp
(keyboard-translate ?h ?s)
(keyboard-translate ?s ?h)
(global-set-key "h" "s")
(global-set-key "s" "h")
```
I get stuck in an infinite loop though...

**Comment by syl20bnr** [comment]
Make it work in the micro-states is hard because they are a special kind of keymap. I can add the possibility to extend a micro-state where bindings could be overridden but it will be painful for you to extend each micro-state.

I like your approach to swapping, it would be really interesting to have it generalized in spacemacs but it is a pretty big development. This is the kind of stuff that's hard to implement when it has not been thought from start :-(

But since we are in emacs and elisp is a so cool language for hacking, we must be able to find a dirty hack :-) I don't know your level in elisp but it might be a good exercise.
The idea is to advice the macro `spacemacs|define-macro-state` [1] with an `around` advice, then you should be able to intercept the arguments and do your swappings before applying your modified arguments to the macro. You can retrieve the passed arguments with `ad-get-arg`.

Tell me if you need help to do this.

[1] https://github.com/syl20bnr/spacemacs/blob/develop/core/core-micro-state.el#L40
[2] https://github.com/syl20bnr/spacemacs/blob/develop/core/core-themes-support.el#L137

**Comment by Nupf** [comment]
Thanks for your consideration :)! 
I guess I was hoping for a quick fix - I'll tackle this when I'm more confident in emacs. That would indeed be a good chance to learn. 
For now I'll integrate what I can from spacemacs into my dotfiles - too little time :/.

Thanks again though :)

**Comment by StreakyCobra** [comment]
@Nupf This is indeed a complicate problem for people like us having a different keyboard layout. I did a contribution layer for the [bepo layout](https://github.com/syl20bnr/spacemacs/blob/develop/layers/+keyboard-layouts/bepo/README.org) that may inspire you. For now all the mappings are changed by hand, but I'm thinking to automatize this a little bit. If you feels motivated, you can start your own for neo-layout layer :smile:

**Comment by a13ph** [comment]
hjkl is a very special case - one of few spacemacs sets of keys which are not mnemonic at all, and rely on positioning (procedural memory) 100%

Therefore I'd propose to modularize them in the long run somehow, or at least make them more customizable on the user side, at the very least by providing documentation section on them

**Comment by AlejandroCatalina** [comment]
I suggest placing the bepo layer functions in the core, because i used pretty much the same functions when i made my dvorak layout layer.

**Comment by a13ph** [comment]
Including it's derivatives. At bare minimum, having a way to list those all would be very helpful.

I wonder if there are any nonderivative of hjkl keys which exhibit same position reliance and lack of mnemonics

Mnemonics part is important there because they don't care about position themselves. Even if it's very good idea to bind often used commands to easy accessed keys, that's not how spacemacs bindings usually work, since we put discoverability first ("hmmm, how do I delete? I'll try `d`") and leave power users ability to remap for their use cases as an override.

*Tangent: vim language is only very mnemonic for english speakers*

**Comment by a13ph** [comment]
Maybe a shared module would be better than direct inclusion into core, I'm not sure. It is an edge case, after all.

**Comment by StreakyCobra** [comment]
The best would be to rename the `keyboard-layouts/bepo` layer to `keyboard-layouts`, and include `dvorak` and `bepo` modifications in it, so that all useful functions would be shared. I already have an idea how to do this cleanly. @AlejandroCatalina Let me know when you `dvorak` layer is ready so I can start working on it.

**Comment by AlejandroCatalina** [comment]
Okay @StreakyCobra I'll PR it later. I had to rename your functions because the layer didn't depend on bepo so when we make it generic we won't have to rename the functions.

On Feb 25, 2016 13:38, Fabien Dubosson <notifications@github.com> wrote:

The best would be to rename the keyboard-layouts/bepo layer to keyboard-layouts, and include dvorak and bepo modifications in it, so that all useful functions would be shared. I already have an idea how to do this cleanly. @AlejandroCatalina Let me know when you dvorak layer is ready so I can start working on it.

**Comment by sindikat** [comment]
It would be great if `keyboard-layouts` layer provided an abstraction (an API) so that for anyone it would be easy to redefine keys however they want, and the changes would propagate to evilified buffers and transient-states automatically. All I'm afraid of is that it's introducing yet another abstraction layer.

Thing to keep in mind: people often have 2 or more keyboard layouts on the same keyboard, e.g. when they use multiple languages. My Russian layout doesn't work in Emacs at all, so it would be great if I could just provide how Russian layout keys correspond to QWERTY (or whatever keys), and all the commands would be bound to Russian layout keys accordingly.

There are physical keyboards keys, there are letters they correspond to (in this or that layout), and there are commands bound to these letters, which are 3 different abstractions.

**Comment by syl20bnr** [comment]
the keyboard-layout layer is now in develop branch.

**Comment by leezu** [comment]
Is there any ongoing work on adding support for the neo-layout @niklasfasching  mentioned to the keyboard-layout layer?

**Comment by d12frosted** [comment]
Fixed with released of Spacemacs v0.200. Let us know if you still have any problems with this issue.
